wgsl/index.bs

```diff
@@ -6981,13 +6979,326 @@ See [[#statements]] and [[#function-calls]].
+      <td>*CFend*
+      <td class="nowrap">(*CF*, *e*) => (*CF'*, *V*)<br>
+        (*V*, *s_1*) => *CF_1*<br>
+        ...<br>
+        if *s_(n-1)* may fallthrough, (*CF_(n-1)*, *s_n*) => *CF_n*<br>
+        else (*V*, *s_n*) => *CF_n*
+      <td>*CFend*
+      <td class="nowrap">*CFend* -> {*CF_1*, ..., *CF_n*}
+  <tr><td class="nowrap">break;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+  <tr><td class="nowrap">continue;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+<tr><td class="nowrap">fallthrough;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+  <tr><td class="nowrap">discard;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+  <tr><td class="nowrap">return;
+      <td>
+      <td>
+      <td>*CF*
+      <td>*CF_return* -> *CF*
+  <tr><td class="nowrap">return *e*;
+      <td>
+      <td class="nowrap">(*CF*, *e*) => (*CF'*, *V*)
+      <td>*CF'*
+      <td>*CF_return* -> *CF'*<br>
+          *Value_return* -> *V*
+  <tr><td class="nowrap">e1 = e2;
+      <td>
+      <td class="nowrap">LValue: (*CF*, *e1*) => (*CF1*, *L1*)<br>
+          (*CF1*, *e2*) => (*CF2*, *V2*)
+      <td>*CF2*
+      <td>*L1* -> *V2*
+</table>
+
+### Uniformity rules for function calls ### {#uniformity-function-calls}
+
+The most complex rule is for function calls:
+- For each argument, apply the corresponding expression rule, with the control flow at the exit of the previous argument (using the control flow at the beginning of the function call for the first argument). Name the corresponding value nodes "arg_i" and the corresponding control flow nodes "CF_i"
+- Create two new nodes, named "Result" and "CF_after"
+- If the control-flow at the start of the function was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to the last CF_i
+- Otherwise add an edge from CF_after to the last CF_i
+- If the function was tagged ControlFlowCannotBeUniformAfter, then add an edge from CF_after to CannotBeUniform
+- Else if it was tagged ReturnValueCannotBeUniform, then add an edge from Result to CannotBeUniform
+- Add an edge from Result to CF_after
+- For each argument *i*:
+    - If the corresponding parameter was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to arg_i
+    - Else if it was tagged MustBeUniformForControlFlowAfter, then add an edge from CF_after to arg_i
+    - Else if it was tagged MustBeUniformForReturnValue, then add an edge from Result to arg_i
+
+Note: Notice that this rule only requires adding a number of edges bounded by 3 + the number of parameters of the functions, independently of how complex the implementation of the function might be. This is key to the linear complexity of the overall algorithm.
+
+Most built-in functions have tags of:
+- for the control-flow in which they can be called: MustBeUniformForControlFlowAfter
+- for the function as a whole: NoRestriction
+- for each parameter: MustBeUniformForReturnValue
+
+Here is the list of exceptions:
```

Comment:
Add:

All functions in Synchronization built-in functions are tagged with AlwaysMustBeUniform.

Reply:
It is already in the list of exceptions just below:
> All functions in [[#texturesample]], [[#texturesamplebias]], [[#texturesamplecompare]], and [[#sync-builtin-functions]] are AlwaysMustBeUniform

Reply:
Oops I missed that. Related question: why are explicit derivatives tagged ReturnValueCannotBeUniform, but functions that depend on implicit derivatives are not? Consider two examples:
```
let dx = dpdx(coord);
let dy = dpdy(coord);
let sample = textureSampleGrad(t, s, vec2<f32>(0.0, 0.0), dx, dy);
if (sample.x > 0.0) {
  textureSample(...);
}
```
```
let sample = textureSample(t, s, vec2<f32>(0.0, 0.0));
if (sample.x > 0.0) {
  textureSample(...);
}
```

It seems like the first example would be rejected and the second would not. The texture and sampler are both read only and the coordinates are constant.

Reply:
Good catch, they should match.
I think they should both have non-uniform return, do you agree?

Reply:
Yes, I think so.